keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -666,7 +773,1310 @@ ways.  One is that we are dispatching requests to be served rather
+
+To make the scheduling technically easy to speccify, we suppose that
+no two requests arrive at the same time.  The implementation will be
+serialized anyway, so this is no real restriction.
+
+```
+B(i,j) = if j = 0 or E(i,j-1) <= t_arrive(i,j)
+         then t_arrive(i,j)
+         else E(i,j-1)
+```
+
+That is, a newly arrived request is dispatched immediately if the
+queue had nothing executing otherwise the new request waits until all
+of the queue's earlier requests finish.  Note that the concurrency
+limit used here is different from the real world: a queue is allowed
+to run one request at a time, regardless of how many it has waiting to
+run and regardless of the server's concurrency limit `C`.  The
+independent virtual-world scheduling for each queue is crucial for
+fairness: a queue's virtual completions depend only on the queue's
+demand and allocated concurrency, not any detailed scheduling
+interaction.  This also helps enable efficient implementations.
+
+The end of a request's virtual execution (`E(i,j)`) is the solution to
+the following equation.
+
+```
+Integral[from tau=B(i,j) to tau=E(i,j)] rate(i,tau) dtau = len(i,j)
+```
+
+That is, each of a queue's requests is executed in the virtual world at
+the aforementioned `rate`.  Before the real completion, `len` is only
+a guess.
+
+For a given request `(i,j)`, the initial guess at its execution
+duration is the sum of `Y(i,j)` and what the request context says is
+the available remaining time for serving the request.  If and whenever
+the passage of time later proves the current guessed `len` to be too
+small then it is increased by adding `G` (which is the server's
+configured default request service time limit).
+
+Once a request `(i,j)` finishes executing in the real world, its
+actual execution duration is known and `len` gets set to that plus
+`Y(i,j)`.  This changes not only `E(i,j)` but also the `B` and `E` of
+the all the queue's requests that arrived between `t_arrive(i,j)` and
+the next of the queue's idle times.  Note that this can change
+`rho(i,t)` and thus `mu_equal(t)` at those intervening times, and thus
+the subsequent scheduling in other queues.  The computation of these
+changes can not happen until `(i,j)` finishes executing in the real
+world (which is `t_arrive(i,j) + len(i,j)`) --- but the request might
+finish much sooner in the virtual world.  We can have `E(i,j) <
+t_arrive(i,j) + len(i,j)` because of `rate` being greater than 1.  An
+accurate implementation would keep track of enough historical
+information to revise all that scheduling.
+
+The order of request dispatches in the real world is taken to be the
+order of request completions in the virtual world.  In the case of
+ties, round robin ordering is used starting from the queue that most
+closely follows the one last dispatched from.  Requests are dispatched
+as soon as allowed by that ordering, the concurrency bound in the
+problem statement, and of course not dispatching before arrival.
```

Comment:
i don't think you need the 'of course..' bit.

Reply:
removed